File: src/main/scala/llmTests/a201802a/java/gpt4/TournamentBuilderImpl.java
package llmTests.a201802a.java.gpt4;


import java.util.*;
import java.util.stream.Collectors;
import llmTests.a201802a.java.Tournament;
import llmTests.a201802a.java.TournamentBuilder;

import llmTests.a201802a.java.Tournament;
import llmTests.a201802a.java.Tournament.Result;
import llmTests.a201802a.java.Tournament.Type;
import llmTests.a201802a.java.TournamentBuilder;

public class TournamentBuilderImpl implements TournamentBuilder {
	private String name;
	private Tournament.Type type;
	private Map<String, Integer> ranking;
	private Map<String, Tournament.Result> results;
	private boolean built = false;

	public TournamentBuilderImpl() {
		this.results = new HashMap<>();
	}

	@Override
	public TournamentBuilder setType(Tournament.Type type) {
		if (type == null || built)
			throw new IllegalStateException();
		this.type = type;
		return this;
	}

	@Override
	public TournamentBuilder setName(String name) {
		if (name == null || built)
			throw new IllegalStateException();
		this.name = name;
		return this;
	}

	@Override
	public TournamentBuilder setPriorRanking(Map<String, Integer> ranking) {
		if (ranking == null || type == null || built)
			throw new IllegalStateException();
		this.ranking = new HashMap<>(ranking);
		return this;
	}

	@Override
	public TournamentBuilder addResult(String player, Tournament.Result result) {
		if (ranking == null || results.containsKey(player) || built)
			throw new IllegalStateException();
		if (result == Tournament.Result.WINNER && results.values().contains(Tournament.Result.WINNER))
			throw new IllegalStateException();
		this.results.put(player, result);
		return this;
	}

	@Override
	public Tournament build() {
		if (name == null || type == null || ranking == null || !results.containsValue(Tournament.Result.WINNER) || built)
			throw new IllegalStateException();
		built = true;
		return new TournamentImpl(name, type, ranking, results);
	}

	private class TournamentImpl implements Tournament {
		private String name;
		private Type type;
		private Map<String, Integer> initialRank;
		private Map<String, Integer> resultingRank;
		private Map<String, Result> results;

		private TournamentImpl(String name, Type type, Map<String, Integer> initialRank, Map<String, Result> results) {
			this.name = name;
			this.type = type;
			this.initialRank = initialRank;
			this.resultingRank = calculateFinalRanking(initialRank, results, type);
			this.results = results;
		}

		private Map<String, Integer> calculateFinalRanking(Map<String, Integer> initialRank, Map<String, Result> results, Type type) {
			Map<String, Integer> finalRank = new HashMap<>();
			int points;
			for(String player : initialRank.keySet()) {
				finalRank.put(player, initialRank.get(player) + pointsAwarded(type, results.getOrDefault(player, Result.PARTICIPANT)));
			}
			for (Map.Entry<String, Result> entry : results.entrySet()) {
				if (!finalRank.containsKey(entry.getKey())) {
					finalRank.put(entry.getKey(), pointsAwarded(type, entry.getValue()));
				}
			}
			return finalRank;
		}

		private int pointsAwarded(Type type, Result result) {
			int points = 0;
			switch (type) {
				case MAJOR: {
					points = 2500;
					break;
				}
				case ATP1000: {
					points = 1000;
					break;
				}
				case ATP500: {
					points = 500;
					break;
				}
				case ATP250: {
					points = 250;
					break;
				}
			}

			switch (result) {
				case FINALIST: {
					points = points/2;
					break;
				}
				case SEMIFINALIST: {
					points = points/5;
					break;
				}
				case QUARTERFINALIST: {
					points = points/10;
					break;
				}
				case PARTICIPANT: {
					points = 0;
					break;
				}
			}
			return points;
		}

		@Override
		public Type getType() {
			return type;
		}

		@Override
		public String getName() {
			return name;
		}

		@Override
		public Optional<Result> getResult(String player) {
			return Optional.ofNullable(results.get(player));
		}

		@Override
		public String winner() {
			return results.entrySet().stream().filter(entry -> entry.getValue() == Result.WINNER).findFirst().get().getKey();
		}

		@Override
		public Map<String, Integer> initialRanking() {
			return initialRank;
		}

		@Override
		public Map<String, Integer> resultingRanking() {
			return resultingRank;
		}

		@Override
		public List<String> rank() {
			return resultingRank.entrySet().stream().sorted(Map.Entry.<String, Integer>comparingByValue().reversed()).map(Map.Entry::getKey).collect(Collectors.toList());
		}
	}
}

File: src/main/scala/llmTests/a201802a/scala/gpt4/TournamentBuilderImpl.scala
//package llmTests.a201802a.scala.gpt4
//import llmTests.a201802a.scala.{Tournament, TournamentBuilder}
//
//import java.util.{Optional, HashMap}
//import scala.jdk.CollectionConverters._
//
//class TournamentBuilderImpl extends TournamentBuilder {
//  private var name: Option[String] = None
//  private var ttype: Option[Tournament.Type] = None
//  private var initialRank: Option[Map[String, Integer]] = None
//  private var results: Map[String, Tournament.Result] = Map.empty
//  private var built: Boolean = false
//
//  override def setType(`type`: Tournament.Type): TournamentBuilder = {
//    require(`type` != null, "Type can't be null!")
//    require(!built, "Tournament has been built!")
//    ttype = Some(`type`)
//    this
//  }
//
//  override def setName(name: String): TournamentBuilder = {
//    require(name != null, "Name can't be null!")
//    require(!built, "Tournament has been built!")
//    this.name = Some(name)
//    this
//  }
//
//  override def setPriorRanking(ranking: java.util.Map[String,Integer]): TournamentBuilder = {
//    require(name.isDefined, "Name has to be defined before setting ranking!")
//    require(ttype.isDefined, "Type has to be defined before setting ranking!")
//    require(!built, "Tournament has been built!")
//    this.initialRank = Some(ranking.asScala.toMap)
//    this
//  }
//
//  override def addResult(player: String, result: Tournament.Result): TournamentBuilder = {
//    require(initialRank.isDefined, "Ranking has to be defined before adding results!")
//    require(!built, "Tournament has been built!")
//    require(!results.contains(player), "Player has already a set result!")
//    require(results.count(_._2 == Tournament.Result.WINNER) == 0 || result != Tournament.Result.WINNER, "There can only be one winner!")
//    results += (player -> result)
//    this
//  }
//
//  override def build(): Tournament = {
//    require(name.isDefined, "Name is mandatory!")
//    require(ttype.isDefined, "Type is mandatory!")
//    require(initialRank.isDefined, "Ranking is mandatory!")
//    require(results.values.count(_ == Tournament.Result.WINNER) == 1, "One player has to be the winner!")
//
//    built = true
//
//    var resultingRank: Map[String, Integer] = initialRank.get
//    results.foreach {
//      case (player, result) => {
//        val playerInitialRank = initialRank.get.getOrElse(player, 0)
//        val points = pointsFor(result, ttype.get)
//        resultingRank = resultingRank.updated(player, Integer.valueOf(playerInitialRank.intValue() + points))
//      }
//    }
//
//    new Tournament {
//      private val finalRank: Map[String,Integer] = resultingRank
//      private val tournamentWinner: String = results.find(_._2 == Tournament.Result.WINNER).get._1
//      override def getType: Tournament.Type = ttype.get
//      override def getName: String = name.get
//      override def initialRanking(): java.util.Map[String,Integer] = initialRank.get.asJava
//      override def getResult(player: String): Optional[Tournament.Result] = Optional.ofNullable(results.getOrElse(player, null))
//      override def resultingRanking(): java.util.Map[String,Integer] = finalRank.asJava
//      override def winner(): String = tournamentWinner
//      override def rank(): java.util.List[String] = finalRank.toSeq.sortBy(-_._2).map(_._1).asJava
//    }
//  }
//
//  private def pointsFor(result: Tournament.Result, `type`: Tournament.Type): Int = {
//    val basePoints = `type` match {
//      case Tournament.Type.MAJOR => 2500
//      case Tournament.Type.ATP1000 => 1000
//      case Tournament.Type.ATP500 => 500
//      case Tournament.Type.ATP250 => 250
//    }
//
//    result match {
//      case Tournament.Result.WINNER => basePoints
//      case Tournament.Result.FINALIST => (basePoints * 0.5).toInt
//      case Tournament.Result.SEMIFINALIST => (basePoints * 0.2).toInt
//      case Tournament.Result.QUARTERFINALIST => (basePoints * 0.1).toInt
//      case Tournament.Result.PARTICIPANT => 0
//    }
//  }
//}